lib/mongo/server/connection_pool.rb

```diff
@@ -307,11 +307,12 @@ def check_out(service_id: nil)
             # a connection while this thread is waiting for one.
             @lock.synchronize do
               until @available_connections.empty?
-                connection = next_available_connection(service_id: service_id)
+                connection = next_available_connection(
+                  connection_global_id: connection_global_id
+                )
 
-                # If service_id is not nil, connection may be nil here
+                # If connection_global_id is not nil, connection may be nil here
                 # even if there are available connections in the pool
-                # (they could be to other services).
                 break unless connection
 
```

Comment:
Unlike with service id where a connection with the wanted service id might spontaneously appear in the future, this cannot happen with connection global ids. Therefore at this point I suggest checking if any connections in the pool have the wanted cgid and if none exist, raise an exception here immediately.

Reply:
Good catch, thanks. Fixed.